src/common/meta/src/ddl/alter_logical_tables.rs

```diff
@@ -116,17 +134,62 @@ impl AlterLogicalTablesProcedure {
             }
         }
 
-        future::join_all(alter_region_tasks)
+        // Collects responses from all the alter region tasks.
+        let phy_raw_schemas = future::join_all(alter_region_tasks)
             .await
             .into_iter()
+            .map(|res| res.map(|mut res| res.extension.remove(ALTER_PHYSICAL_EXTENSION_KEY)))
             .collect::<Result<Vec<_>>>()?;
 
-        self.data.state = AlterTablesState::UpdateMetadata;
+        if phy_raw_schemas.is_empty() {
+            self.data.state = AlterTablesState::UpdateMetadata;
+            return Ok(Status::executing(true));
+        }
+
+        // Verify all the physical schemas are the same
+        // Safety: previous check ensures this vec is not empty
+        let first = phy_raw_schemas.first().unwrap();
+        ensure!(
+            phy_raw_schemas.iter().all(|x| x == first),
+            MetadataCorruptionSnafu {
+                err_msg: "The physical schemas from datanodes are not the same."
+            }
+        );
 
+        // Decodes the physical raw schemas
+        if let Some(phy_raw_schema) = first {
+            self.data.physical_columns =
+                ColumnMetadata::decode_list(phy_raw_schema).context(DecodeJsonSnafu)?;
+        } else {
+            warn!("altering logical table result doesn't contains extension key `{ALTER_PHYSICAL_EXTENSION_KEY}`,leaving the physical table's schema unchanged");
+        }
+
+        self.data.state = AlterTablesState::UpdateMetadata;
         Ok(Status::executing(true))
     }
 
     pub(crate) async fn on_update_metadata(&mut self) -> Result<Status> {
+        if !self.data.physical_columns.is_empty() {
+            let physical_columns = std::mem::take(&mut self.data.physical_columns);
+            let physical_table_info = self.data.physical_table_info.as_ref().unwrap();
+
+            // Generates new table info
+            let old_raw_table_info = physical_table_info.table_info.clone();
+            let new_raw_table_info = physical_table_metadata::build_new_physical_table_info(
+                old_raw_table_info,
+                physical_columns,
+            );
+
+            // Updates physical table's metadata
+            self.context
+                .table_metadata_manager
+                .update_table_info(
+                    DeserializedValueWithBytes::from_inner(physical_table_info.clone()),
+                    new_raw_table_info,
+                )
+                .await?;
```

Comment:
If this method returns an error, then `self.data.physical_columns` is always empty when we retry the procedure.